### PR TITLE
 Decoupling authentication from launch in the SalesforceSDKManager

### PR DESCRIPTION
**Pull request by trooper2013**
Deprecated launch related functions. Added 2 net new Api(s) to the SDKManager. Changed code in the AppDelegate's for samples. Now the only thing required is to invoke the initializeSDK for the Manager. Since the process of init merely set's up configurations  for the sdk in a synchronous fashion there is no need for setting up  launch  or authentication related blocks. The developer has to invoke the SFUserAccountManager login api(s) on-demand. 

Bootstrap the SDK using the following.
```
[SalesforceSDKManager initializeSDK];  
[SalesforceSDKManager initializeSDKWithClass:SomeClass.class];

or using any  one of relevant subclasses

[SmartSyncSDKManager initializeSDK];
[SalesforceHybridSDKManager initializeSDK];
```

For Logout & User Switch Notifications 
```
[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleUserDidLogout:) name:kSFNotificationUserDidLogout object:nil];
        
[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleUserDidSwitch:) name:kSFNotificationUserDidSwitch  object:nil];
```

### CODE REVIEW

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SalesforceHybridSDKManager.m [hunk 1]
```diff
@@ -41,6 +41,10 @@ - (NSString*) pathForUserSyncsConfig;
 @implementation SalesforceHybridSDKManager
 @synthesize appConfig = _appConfig;
 
++ (void)initializeSDK {
```
Comment (bhariharan) [inline libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SalesforceHybridSDKManager.m]: Should we just call this `init`? The `SDK` part is redundant.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.h [hunk 2]
```diff
@@ -251,7 +251,17 @@ typedef void (^SFSnapshotViewControllerDismissalBlock)(UIViewController* snapsho
  actions.
  @return YES if the launch successfully kicks off, NO if launch is already running.
  */
-- (BOOL)launch;
+- (BOOL)launch SFSDK_DEPRECATED(7.0, 8.0, "SalesforceSDKManager launch has been deprecated. Use initializeSDK instead of launch");
+
+/**
+ Initializes the SDK.
+ */
++ (void)initializeSDK;
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.h]: Should we just call this `init`? The `SDK` part is redundant.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.h [hunk 3]
```diff
@@ -251,7 +251,17 @@ typedef void (^SFSnapshotViewControllerDismissalBlock)(UIViewController* snapsho
...
+
+/**
+ Initializes the SDK.  Class instance to be used to instantiate the sdkManager.
+ */
++ (void)initializeSDKWithClass:(Class)className;
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.h]: `initWithClass`?

File: libs/SmartStore/SmartStore/Classes/SmartStoreSDKManager.m [hunk 4]
```diff
@@ -41,6 +41,10 @@ -(instancetype)init {
     return self;
 }
 
++ (void)initializeSDK {
```
Comment (bhariharan) [inline libs/SmartStore/SmartStore/Classes/SmartStoreSDKManager.m]: Should we just call this `init`? The `SDK` part is redundant.

File: libs/SmartSync/SmartSync/Classes/Manager/SmartSyncSDKManager.m [hunk 5]
```diff
@@ -28,6 +28,10 @@
 
 @implementation SmartSyncSDKManager
 
++ (void)initializeSDK {
```
Comment (bhariharan) [inline libs/SmartSync/SmartSync/Classes/Manager/SmartSyncSDKManager.m]: Should we just call this `init`? The `SDK` part is redundant.

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SalesforceHybridSDKManager.m (hunk 1 above)
Comment (wmathurin) [reply libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SalesforceHybridSDKManager.m]: In objective-c, init is the instance level initializer (constructor) and initialize the class level initializer. So I don't think we should call it init. Having the name based on initialize makes sense.

Comment (bhariharan) [reply libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SalesforceHybridSDKManager.m]: Ah yes. I completely forgot. That makes sense.

File: native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/AppDelegate.m [hunk 6]
```diff
@@ -66,14 +66,12 @@ - (id)init
...
-        [SmartSyncSDKManager sharedManager].appConfig.remoteAccessConsumerKey = config.remoteAccessConsumerKey;
-        [SmartSyncSDKManager sharedManager].appConfig.oauthRedirectURI = config.oauthRedirectURI;
-        [SmartSyncSDKManager sharedManager].appConfig.oauthScopes = [NSSet setWithArray:config.oauthScopes];
-        __weak typeof(self) weakSelf = self;
-        [[SmartSyncSDKManager sharedManager] addDelegate:self];
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleUserDidLogout:) name:kSFNotificationUserDidLogout object:nil];
         
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleUserDidSwitch:) name:kSFNotificationUserDidSwitch  object:nil];
+        [SmartSyncSDKManager initializeSDK];
+        [[SmartSyncSDKManager sharedManager] addDelegate:self];
```
Comment (wmathurin) [inline native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/AppDelegate.m]: Why do we still need to be a delegate of the SDK manager? Couldn't we do everything with notification?

File: native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift [hunk 7]
```diff
@@ -139,6 +129,20 @@ class AppDelegate : UIResponder, UIApplicationDelegate
     }
     
     // MARK: - Private methods
+    func loginIfRequired() -> Void  {
+        if SFUserAccountManager.sharedInstance().currentUser != nil  {
+            self.setupRootViewController()
+        }
+        
+        _ = SFUserAccountManager.sharedInstance().Promises.login()
```
Comment (wmathurin) [inline native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift]: Should that be in the else block?

File: native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift [hunk 8]
```diff
@@ -139,6 +129,20 @@ class AppDelegate : UIResponder, UIApplicationDelegate
     }
     
     // MARK: - Private methods
+    func loginIfRequired() -> Void  {
```
Comment (wmathurin) [inline native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift]: There is still a lot of boilerplate code (loginIfRequired, all the handle* methods). Could we provide that through a helper class and only require a one liner in the AppDelegate?

File: native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/AppDelegate.m (hunk 6 above)
Comment (trooper2013) [reply native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/AppDelegate.m]: @wmathurin the delegate was already  part of  the sample to clear keyboard on backgrounding

File: native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift (hunk 7 above)
Comment (trooper2013) [reply native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift]: Good catch. It should have been.

File: native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/AppDelegate.m (hunk 6 above)
Comment (wmathurin) [reply native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/AppDelegate.m]: Ah ok.

File: native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift (hunk 8 above)
Comment (wmathurin) [reply native/SampleApps/RestAPIExplorer/RestAPIExplorer/AppDelegate.swift]: The only code that are app specific is the setup of the root view controller, and the bit with the initial view controller. Could we move everything to a helper (loginIfRequired, handle logout and user switching)? Of course some apps might need more control and could look at the code of the helper as a guide, but all the sample apps / template apps would have much simple AppDelegate classes.

**Pull request merged**